### PR TITLE
Doc alias name restriction

### LINKED ISSUE #76705: Clean up doc_alias feature
From this comment: https://github.com/rust-lang/rust/issues/50146#issuecomment-686726591

Still remaining the `.replace()`.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #76705.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
cc @Mark-Simulacrum (since you commented on the issue ;) )

r? @ollie27

**Comment by pickfire** [comment]
Should be show the list of stuff not allowed in the name?

File: src/doc/rustdoc/src/advanced-features.md [hunk 1]
```diff
@@ -46,3 +46,6 @@ pub struct BigX;
 
 Then, when looking for it through the `rustdoc` search, if you enter "x" or
 "big", search will show the `BigX` struct first.
+
+There are some limitations on the doc alias names though: you can't use `"` or whitespaces or
+tabs or backlines.
```
Comment (jyn514) [inline src/doc/rustdoc/src/advanced-features.md:51]: I think by 'backlines' you meant 'newlines', but both that and spaces are whitespace.
```suggestion
There are some limitations on the doc alias names though: you can't use `"` or whitespace.
```

File: src/test/rustdoc-ui/check-doc-alias-attr.rs [hunk 2]
```diff
@@ -7,4 +7,8 @@ pub struct Bar;
 #[doc(alias)] //~ ERROR
 #[doc(alias = 0)] //~ ERROR
 #[doc(alias("bar"))] //~ ERROR
+#[doc(alias = "\"")] //~ ERROR
+#[doc(alias = "\n")] //~ ERROR
+#[doc(alias = " ")] //~ ERROR
+#[doc(alias = "\t")] //~ ERROR
```
Comment (jyn514) [inline src/test/rustdoc-ui/check-doc-alias-attr.rs:13]:
```suggestion
#[doc(alias = "\t")] //~ ERROR
#[doc(alias = "
")] //~ ERROR
```

File: src/doc/rustdoc/src/advanced-features.md (hunk 1 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/advanced-features.md:51]: Oh good point.

**Comment by GuillaumeGomez** [comment]
Updated!

File: compiler/rustc_passes/src/check_attr.rs [hunk 3]
```diff
@@ -252,23 +252,43 @@ impl CheckAttrVisitor<'tcx> {
...
+                        }
+                        let doc_alias =
+                            meta.value_str().map(|s| s.to_string()).unwrap_or_else(String::new);
+                        if doc_alias.is_empty() {
+                            self.doc_alias_str_error(meta);
+                            return false;
+                        }
+                        if let Some(c) = doc_alias
+                            .chars()
+                            .find(|&c| c == '"' || c == ' ' || c == '\t' || c == '\r' || c == '\n')
```
Comment (oli-obk) [inline compiler/rustc_passes/src/check_attr.rs:282]:
```suggestion
                            .find(|&c| c == '"' || c.is_whitespace())
```

**Comment by oli-obk** [comment]
Just a nit. r=me with that

**Comment by ollie27** [comment]
Single quotes can also be used for exact search: https://github.com/rust-lang/rust/blob/a585aaefb662ae26f8675955f26fad9be4fea96b/src/librustdoc/html/static/main.js#L1077

Assuming that's the reason to disallow double quotes then surely we shouldn't allow single quotes as well?

If you search for an alias surrounded by quotes: https://doc.rust-lang.org/nightly/std/?search=%22reduce%22 then the alias doesn't show up in the results. Is that a bug?

**Comment by GuillaumeGomez** [comment]
@ollie27: I wrote this code and yet I didn't remember that... Not very serious of me haha. Great catch, thanks!

@bors: r-

**Comment by lopopolo** [comment]
This issue doesn't say why the restriction on whitespace characters was added, but I'd like to share why I was using whitespace characters in doc aliases.

My code has a public `Mt` struct. `Mt` is a random number generator that implements the MT19937 mersenne twister. I want to alias this struct to `mersenne twister` and `Mersenne twister`.

**Pull request merged**